torch/csrc/jit/tensorexpr/mem_arena.cpp

```diff
@@ -5,51 +5,40 @@ namespace torch {
 namespace jit {
 namespace tensorexpr {
 
+thread_local KernelArena* current_arena = nullptr;
+
 KernelArena::~KernelArena() {
   for (KernelScopedObject* p : kernel_objects_) {
     delete p;
   }
 }
 
 KernelScopedObject::KernelScopedObject() {
-  KernelArena& kernel = KernelArena::GetCurrentKernelArena();
-  kernel.kernel_objects_.push_back(this);
+  KernelArena* kernel = KernelArena::GetCurrentKernelArena();
+  kernel->kernel_objects_.push_back(this);
 }
 
 static std::vector<KernelArena*>& GetKernelArenaStack() {
   thread_local std::vector<KernelArena*> kernel_arena_stack;
   return kernel_arena_stack;
 }
 
-KernelArena& KernelArena::GetCurrentKernelArena() {
-  std::vector<KernelArena*>& kernel_arena_stack = GetKernelArenaStack();
-  if (kernel_arena_stack.empty()) {
-    throw std::runtime_error(
-        "A KernelScope must be bound before creating KernelScopedObject");
-  }
-  return *kernel_arena_stack.back();
+void KernelArena::SetCurrentKernelArena(KernelArena *new_kernel_arena) {
+  current_arena = new_kernel_arena;
 }
 
-KernelScope::KernelScope() : owning_kernel_arena_(true) {
-  kernel_arena_ = new KernelArena;
-  GetKernelArenaStack().push_back(kernel_arena_);
+KernelArena* KernelArena::GetCurrentKernelArena() {
+  return current_arena;
 }
 
-KernelScope::KernelScope(KernelArena& kernel_arena)
-    : owning_kernel_arena_(false) {
-  kernel_arena_ = &kernel_arena;
-  GetKernelArenaStack().push_back(&kernel_arena);
+KernelScope::KernelScope() {
+  old_kernel_arena_ = KernelArena::GetCurrentKernelArena();
+  KernelArena::SetCurrentKernelArena(new KernelArena);
 }
 
-KernelScope::~KernelScope() noexcept(false) {
-  std::vector<KernelArena*>& kernel_arena_stack = GetKernelArenaStack();
-  if (kernel_arena_ != kernel_arena_stack.back()) {
-    throw std::runtime_error("Mismatch KernelScope and kernel");
-  }
-  if (owning_kernel_arena_) {
-    delete kernel_arena_;
-  }
-  kernel_arena_stack.pop_back();
+KernelScope::~KernelScope() {
+  delete KernelArena::GetCurrentKernelArena();
```

Comment:
You still need to check for the case where your Arena is not the current Arena. This could be triggered by an illegal case: 

create scope A
create scope B
delete scope A
delete scope B

Reply:
This can only occur if we would allocate scopes dynamically, which is not an intended use anyway: if scope objects are only created as local variables this never happens. While I agree that it'd be nice to check for that situation, doing that in a destructor is a UB, so it doesn't really help.